pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/util/SchemaUtil.java

```diff
@@ -88,8 +90,10 @@ public static Schema extractAvroSchema(SchemaDefinition schemaDefinition, Class
         try {
             return parseAvroSchema(pojo.getDeclaredField("SCHEMA$").get(null).toString());
         } catch (NoSuchFieldException | IllegalAccessException | IllegalArgumentException ignored) {
-            return schemaDefinition.getAlwaysAllowNull() ? ReflectData.AllowNull.get().getSchema(pojo)
-                    : ReflectData.get().getSchema(pojo);
+            ReflectData reflectData = schemaDefinition.getAlwaysAllowNull() ? ReflectData.AllowNull.get()
+                    : ReflectData.get();
+            AvroSchema.addLogicalTypeConversions(reflectData, schemaDefinition.isJsr310ConversionEnabled());
```

Comment:
Can you explain why do we need this change?

Reply:
This code is executed when you build the schema from a POJO that is not generated by Avro.

I guess that without this change the test is not passing

Reply:
But we have registered the logic types in the static block. Why didn't it work?

Reply:
Because there is in fact 2 singleton instances in the AVRO ReflectData class: the ReflectData.INSTANCE and the ReflectData.AllowNull.INSTANCE. So when your schema allows nulls, you need to register logical types conversion in the ReflectData.AllowNull.INSTANCE. That is already done like that in https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/StructSchema.java#L96

Reply:
@vroyer 
I share @sijie's concern.
I didn't notice that we are modifying a singleton.
we could initialise it only once 

but I cannot understand how it can work with `schemaDefinition.isJsr310ConversionEnabled())`, that depends on a particular schema.
if we have two schemas with different values for isJsr310ConversionEnabled, we will have a race and it may happen that we use the wrong configuration (redarding isJsr310ConversionEnabled)

Reply:
Yes, that is a potential issue, but you also have it with the [AvroReader](https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/reader/AvroReader.java#L49), you can have several schemas sharing the same classloader. We can fix that by creating a ReflectData instance each time rather than using the singleton instance ?

Reply:
what about using two instances of ReflectData (in total 4), one for isJsr310ConversionEnabled=true and one for isJsr310ConversionEnabled=false ?

I am not sure if it a problem to share the same ReflectData among more classloaders

Reply:
looks good to me now

@sijie how does it look to your ?

Reply:
Fixed with a new ReflectData instance rather than using a singleton.